test/rtl/canonicalization.mlir

```diff
@@ -1,54 +1,113 @@
   %1 = rtl.or  %0 : i11
   %2 = rtl.xor %1 : i11
   %3 = rtl.add %2 : i11
   %4 = rtl.mul %3 : i11
-  // CHECK-NEXT:    return %arg0
   return %4 : i11
 }
 
 // CHECK-LABEL: func @and_annulment(%arg0: i11, %arg1: i11) -> i11 {
+// CHECK-NEXT:    %c0_i11 = rtl.constant(0 : i11)
+// CHECK-NEXT:    return %c0_i11
+
 func @and_annulment(%arg0: i11, %arg1: i11) -> i11 {
   %c0_i11 = rtl.constant(0 : i11) : i11
   %0 = rtl.and %arg0, %arg1, %c0_i11 : i11
-  // CHECK-NEXT: rtl.constant(0 : i11)
   return %0 : i11
 }
 
 // CHECK-LABEL: func @or_annulment(%arg0: i11) -> i11 {
+// CHECK-NEXT:    %c-1_i11 = rtl.constant(-1 : i11)
+// CHECK-NEXT:    return %c-1_i11
+
 func @or_annulment(%arg0: i11) -> i11 {
   %c-1_i11 = rtl.constant(-1 : i11) : i11
   %0 = rtl.or %arg0, %arg0, %arg0, %c-1_i11 : i11
-  // CHECK-NEXT: rtl.constant(-1 : i11)
   return %0 : i11
 }
 
 // CHECK-LABEL: func @mul_annulment(%arg0: i11, %arg1: i11, %arg2: i11) -> i11 {
+// CHECK-NEXT:    %c0_i11 = rtl.constant(0 : i11) : i11
+// CHECK-NEXT:    return %c0_i11
+
 func @mul_annulment(%arg0: i11, %arg1: i11, %arg2: i11) -> i11 {
-  // CHECK-NEXT: %c0_i11 = rtl.constant(0 : i11) : i11
   %c0_i11 = rtl.constant(0 : i11) : i11
-  %0 = rtl.mul %arg0, %arg1, %c0_i11 : i11
-  // CHECK-NEXT: %0 = rtl.or %arg2, %c0_i11
-  %1 = rtl.or %c0_i11, %arg2: i11
-  // CHECK-NEXT: return %0
-  return %1 : i11
+  %0 = rtl.mul %arg0, %c0_i11, %arg1 : i11
+  return %0 : i11
+}
+
+// Identities
+
+// CHECK-LABEL: func @and_identity(%arg0: i11, %arg1: i11) -> i11 {
+// CHECK-NEXT:    %0 = rtl.and %arg0, %arg1
+// CHECK-NEXT:    return %0
+
+func @and_identity(%arg0: i11, %arg1: i11) -> i11 {
+  %c-1_i11 = rtl.constant(-1 : i11) : i11
+  %0 = rtl.and %c-1_i11, %arg0, %arg1, %c-1_i11 : i11
+  return %0 : i11
+}
+
+// CHECK-LABEL: func @or_identity(%arg0: i11, %arg1: i11) -> i11 {
+// CHECK-NEXT:    %0 = rtl.or %arg0, %arg1
+// CHECK-NEXT:    return %0
+
+func @or_identity(%arg0: i11, %arg1: i11) -> i11 {
+  %c0_i11 = rtl.constant(0 : i11) : i11
+  %0 = rtl.or %arg0, %c0_i11, %arg1 : i11
+  return %0 : i11
+}
+
+// CHECK-LABEL: func @xor_identity(%arg0: i11, %arg1: i11) -> i11 {
+// CHECK-NEXT:    %0 = rtl.xor %arg0, %arg1
+// CHECK-NEXT:    return %0
+
+func @xor_identity(%arg0: i11, %arg1: i11) -> i11 {
+  %c0_i11 = rtl.constant(0 : i11) : i11
+  %0 = rtl.xor %arg0, %c0_i11, %arg1 : i11
+  return %0 : i11
+}
+
+// CHECK-LABEL: func @add_identity(%arg0: i11, %arg1: i11) -> i11 {
+// CHECK-NEXT:    %0 = rtl.add %arg0, %arg1
+// CHECK-NEXT:    return %0
+
+func @add_identity(%arg0: i11, %arg1: i11) -> i11 {
+  %c0_i11 = rtl.constant(0 : i11) : i11
+  %0 = rtl.add %arg0, %c0_i11, %arg1 : i11
+  return %0 : i11
+}
+
+// CHECK-LABEL: func @mul_identity(%arg0: i11, %arg1: i11) -> i11 {
+// CHECK-NEXT:    %0 = rtl.mul %arg0, %arg1
+// CHECK-NEXT:    return %0
+
+func @mul_identity(%arg0: i11, %arg1: i11) -> i11 {
+  %c1_i11 = rtl.constant(1 : i11) : i11
+  %0 = rtl.mul %arg0, %c1_i11, %arg1 : i11
```

Comment:
nice.  Can you add one of these checks with the constant on the left hand side?  The generic canonicalization stuff should move the constant to the right so we should handle that, but we should add one test for this.

Reply:
modified the test